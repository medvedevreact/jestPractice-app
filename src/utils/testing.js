import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

export const renderWithRouter = (component, options = {}) => {
  const Wrapper = ({ children }) => (
    <MemoryRouter {...options}>{children}</MemoryRouter>
  );
  return render(component, { wrapper: Wrapper });
};
