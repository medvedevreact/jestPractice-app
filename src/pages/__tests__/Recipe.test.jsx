import { renderWithRouter } from '../../utils/testing';
import { Recipe } from '../Recipe';
import * as api from '../../api';
import { waitForElementToBeRemoved, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useNavigate } from 'react-router-dom';

const apiSpy = jest.spyOn(api, 'getMealById');

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

const useNavigateMock = jest.mocked(useNavigate);

describe('Recipe', () => {
  test('render', async () => {
    apiSpy.mockResolvedValueOnce({
      meals: [
        {
          idMeal: '1',
          strMeal: 'First',
          strCategory: 'Category',
          strArea: 'Area',
          strInstructions: 'Instructions',
          strMealThumb: 'Thumb',
          strIngridients: 'Ing1',
          strMeasure: 'Measure1',
          strYoutube: 'Youtube',
        },
      ],
    });
    renderWithRouter(<Recipe />);
    const preloader = screen.getByRole('progressbar');
    expect(preloader).toBeInTheDocument();
    await waitForElementToBeRemoved(preloader);
    expect(screen.getByRole('article')).toMatchSnapshot();
  });
  test('navigates back when "Go Back" button is clicked', async () => {
    const navigateMock = jest.fn();

    useNavigateMock.mockReturnValue(navigateMock);

    apiSpy.mockResolvedValueOnce({
      meals: [
        {
          idMeal: '1',
          strMeal: 'First',
          strCategory: 'Category',
          strArea: 'Area',
          strInstructions: 'Instructions',
          strMealThumb: 'Thumb',
          strIngridients: 'Ing1',
          strMeasure: 'Measure1',
          strYoutube: 'Youtube',
        },
      ],
    });

    renderWithRouter(<Recipe />);

    const preloader = screen.getByRole('progressbar');
    await waitForElementToBeRemoved(preloader);

    const goBackButton = screen.getByRole('button', { name: /go back/i });

    await userEvent.click(goBackButton);

    expect(navigateMock).toHaveBeenCalledWith(-1);
    expect(navigateMock).toHaveBeenCalledTimes(1);
  });
});
