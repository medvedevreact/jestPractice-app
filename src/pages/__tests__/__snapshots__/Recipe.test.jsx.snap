// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe render 1`] = `
<div
  class="recipe"
  role="article"
>
  <img
    alt="First"
    src="Thumb"
  />
  <h1>
    First
  </h1>
  <h6>
    Category: 
    Category
  </h6>
  <h6>
    Area: 
    Area
  </h6>
  <p>
    Instructions
  </p>
  <table
    class="centered"
  >
    <thead>
      <tr>
        <th>
          Ingredient
        </th>
        <th>
          Measure
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <div
    class="row"
  >
    <h5
      style="margin: 2rem 0px 1.5rem;"
    >
      Video Recipe
    </h5>
    <iframe
      allowfullscreen=""
      src="https://www.youtube.com/embed/Youtube"
    />
  </div>
</div>
`;
